File: apps/android/app/src/main/kotlin/com/loonext/android/features/contacts/DuplicateContactsCard.kt
package com.loonext.android.features.contacts

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.selection.selectable
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.RadioButton
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.dp
import com.loonext.android.core.model.ContactMergeResult
import com.loonext.android.core.model.DuplicatePair
import com.loonext.android.ui.common.formatPhone
import com.loonext.android.ui.common.rememberHaptics
import com.loonext.android.ui.common.userMessage
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.launch

/**
 * #246 — the duplicates this workspace has, offered rather than hunted for.
 *
 * # Why it sits above the contact list and not behind its own screen
 *
 * "The workspace can find its likely duplicates without knowing they exist."
 * Somebody who does not know they have duplicates will not navigate to a screen
 * about them. The card appears above the list only when there is something to
 * act on, which makes it a finding rather than a feature — a workspace with no
 * duplicates sees the list exactly as it always did.
 *
 * *Applying: Meaningful Highlights & Context — the pair IS the insight, so it
 * is one line each with the server's reason attached. Zen of Clarity — no card
 * at all when there is nothing to merge.*
 */
@Composable
fun DuplicateContactsCard(
    repo: ContactMutations,
    companyId: String,
    /** #246: merging needs settings.manage. A member still sees the finding. */
    canMerge: Boolean,
    /**
     * Handed the result rather than a bare signal, because the opt-out union is
     * the one outcome the crew has to be told about — a merged contact can come
     * out opted out when neither record they were looking at said so.
     */
    onMerged: (ContactMergeResult) -> Unit,
    modifier: Modifier = Modifier,
) {
    var pairs by remember { mutableStateOf<List<DuplicatePair>>(emptyList()) }
    var refreshKey by remember { mutableIntStateOf(0) }
    var merging by remember { mutableStateOf<DuplicatePair?>(null) }

    LaunchedEffect(companyId, refreshKey) {
        pairs = try {
            repo.duplicates(companyId).data
        } catch (cause: CancellationException) {
            throw cause
        } catch (_: Exception) {
            // A finding nobody asked for must never become an error somebody
            // has to dismiss. Silence is the honest failure here.
            emptyList()
        }
    }

    if (pairs.isEmpty()) return

    Surface(
        color = MaterialTheme.colorScheme.surface,
        modifier = modifier.fillMaxWidth().padding(horizontal = 16.dp, vertical = 8.dp),
    ) {
        Column(Modifier.padding(14.dp)) {
            Text(
                if (pairs.size == 1) "These two look like the same customer"
                else "${pairs.size} pairs look like the same customer",
                style = MaterialTheme.typography.titleSmall,
            )
            Text(
                "Merging keeps every message, task and photo from both, under one record.",
                style = MaterialTheme.typography.bodySmall,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                modifier = Modifier.padding(top = 2.dp),
            )
            pairs.forEach { pair ->
                Spacer(Modifier.height(10.dp))
                Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
                    Column(Modifier.weight(1f)) {
                        Text(
                            describeContact(pair.name_a, pair.phone_a) +
                                " and " + describeContact(pair.name_b, pair.phone_b),
                            style = MaterialTheme.typography.bodyMedium,
                            maxLines = 2,
                            overflow = TextOverflow.Ellipsis,
                        )
                        // The reason, in the words the server used. A suggestion
                        // somebody cannot verify is one they learn to dismiss.
                        Text(
                            pair.reason,
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                        )
                    }
                    if (canMerge) {
                        TextButton(onClick = { merging = pair }) { Text("Merge") }
                    }
                }
            }
        }
    }

    merging?.let { pair ->
        MergeContactsDialog(
            repo = repo,
            companyId = companyId,
            pair = pair,
            onDismiss = { merging = null },
            onMerged = { result ->
                merging = null
                refreshKey++
                onMerged(result)
            },
        )
    }
}

/** A contact as somebody recognises it: the name if there is one, else the number. */
internal fun describeContact(name: String?, phone: String): String {
    val trimmed = name?.trim()
    return if (trimmed.isNullOrEmpty()) formatPhone(phone)
    else "$trimmed (${formatPhone(phone)})"
}

/**
 * # Ethical Friction, and which direction the dialog states
 *
 * A merge moves somebody's whole history under a different record. The undo
 * restores the second contact but NOT which thread came from which, so this
 * says out loud what survives and names the direction in the way people get
 * backwards ("merge A into B" is ambiguous to almost everyone).
 *
 * Both numbers keep working either way — the fact that makes the decision safe
 * and the one most likely to be assumed wrong.
 */
@Composable
private fun MergeContactsDialog(
    repo: ContactMutations,
    companyId: String,
    pair: DuplicatePair,
    onDismiss: () -> Unit,
    onMerged: (ContactMergeResult) -> Unit,
) {
    var keepFirst by remember(pair) { mutableStateOf(true) }
    var saving by remember { mutableStateOf(false) }
    var error by remember { mutableStateOf<String?>(null) }
    val coroutines = rememberCoroutineScope()
    val haptics = rememberHaptics()

    val survivorId = if (keepFirst) pair.contact_a else pair.contact_b
    val foldedId = if (keepFirst) pair.contact_b else pair.contact_a
    val survivorLabel =
        if (keepFirst) describeContact(pair.name_a, pair.phone_a)
        else describeContact(pair.name_b, pair.phone_b)
    val foldedLabel =
        if (keepFirst) describeContact(pair.name_b, pair.phone_b)
        else describeContact(pair.name_a, pair.phone_a)

    AlertDialog(
        onDismissRequest = { if (!saving) onDismiss() },
        title = { Text("Merge these two customers") },
        text = {
            Column {
                Text(
                    "Everything from both — messages, tasks, photos, notes — ends up " +
                        "under the record you keep. Both phone numbers keep working.",
                    style = MaterialTheme.typography.bodyMedium,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                Spacer(Modifier.height(10.dp))
                Text("Which one to keep", style = MaterialTheme.typography.labelLarge)
                listOf(true, false).forEach { first ->
                    val label =
                        if (first) describeContact(pair.name_a, pair.phone_a)
                        else describeContact(pair.name_b, pair.phone_b)
                    Row(
                        Modifier
                            .fillMaxWidth()
                            .selectable(
                                selected = keepFirst == first,
                                enabled = !saving,
                                onClick = { keepFirst = first },
                            )
                            .padding(vertical = 4.dp),
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(4.dp),
                    ) {
                        RadioButton(
                            selected = keepFirst == first,
                            onClick = { keepFirst = first },
                            enabled = !saving,
                        )
                        Text(label, style = MaterialTheme.typography.bodyMedium)
                    }
                }
                Spacer(Modifier.height(6.dp))
                // Said back in the direction people get backwards.
                Text(
                    "$foldedLabel stops being a separate customer. Its history moves " +
                        "to $survivorLabel.",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant,
                )
                error?.let {
                    Spacer(Modifier.height(6.dp))
                    Text(
                        it,
                        style = MaterialTheme.typography.bodySmall,
                        color = MaterialTheme.colorScheme.error,
                    )
                }
            }
        },
        confirmButton = {
            TextButton(
                enabled = !saving,
                onClick = {
                    haptics.confirm()
                    error = null
                    saving = true
                    coroutines.launch {
                        try {
                            onMerged(repo.merge(companyId, foldedId, survivorId))
                        } catch (cause: Exception) {
                            error = cause.userMessage()
                        } finally {
                            saving = false
                        }
                    }
                },
            ) { Text(if (saving) "Merging…" else "Merge") }
        },
        dismissButton = {
            TextButton(enabled = !saving, onClick = onDismiss) { Text("Cancel") }
        },
    )
}

File: apps/ios/Loonext/Features/Contacts/DuplicateContactsCard.swift
import SwiftUI

/// #246 — the duplicates this workspace has, offered rather than hunted for.
///
/// # Why it sits above the contact list and not behind its own screen
///
/// "The workspace can find its likely duplicates without knowing they exist."
/// Somebody who does not know they have duplicates will not navigate to a
/// screen about them. The card appears above the list only when there is
/// something to act on, which makes it a finding rather than a feature — a
/// workspace with no duplicates sees the list exactly as it always did.
///
/// *Applying: Meaningful Highlights & Context — the pair IS the insight, so it
/// is one line each with the server's reason attached. Zen of Clarity — no card
/// at all when there is nothing to merge.*
@MainActor
struct DuplicateContactsCard: View {
    let mutations: ContactMutations
    let companyId: String
    /// #246: merging needs settings.manage. A member still sees the finding.
    let canMerge: Bool
    /// Handed the result rather than a bare signal, because the opt-out union
    /// is the one outcome the crew has to be told about — a merged contact can
    /// come out opted out when neither record they were looking at said so.
    let onMerged: @MainActor (ContactMergeResult) -> Void

    @State private var pairs: [DuplicatePair] = []
    @State private var refreshKey = 0
    @State private var merging: DuplicatePair?

    var body: some View {
        // The card is absent, not empty: no skeleton and no "you have no
        // duplicates" state. Both would be a screenful of nothing for the
        // workspaces this feature has no findings for, which is most of them.
        //
        // A VStack rather than a Group so the container is in the hierarchy
        // even with nothing to show — the fetch below hangs off it, and a
        // modifier on a view that resolved to EmptyView is not reliably run.
        VStack(spacing: 0) {
            if !pairs.isEmpty {
                PaperCard {
                    VStack(alignment: .leading, spacing: 0) {
                        header
                        ForEach(pairs) { pair in
                            RowDivider()
                            row(pair)
                        }
                    }
                }
                .padding(.horizontal, 18)
                .padding(.top, 6)
            }
        }
        .task(id: "\(companyId)|\(refreshKey)") {
            do {
                pairs = try await mutations.duplicates(companyId: companyId).data
            } catch {
                // A finding nobody asked for must never become an error
                // somebody has to dismiss. Silence is the honest failure here.
                pairs = []
            }
        }
        .sheet(item: $merging) { pair in
            MergeContactsSheet(
                mutations: mutations,
                companyId: companyId,
                pair: pair,
                onMerged: { result in
                    merging = nil
                    refreshKey += 1
                    onMerged(result)
                }
            )
        }
    }

    private var header: some View {
        HStack(alignment: .top, spacing: 10) {
            Image(systemName: "person.2")
                .font(.system(size: 14, weight: .medium))
                .foregroundStyle(BrandColor.muted500)
                .padding(.top, 1)
            VStack(alignment: .leading, spacing: 2) {
                Text(
                    pairs.count == 1
                        ? "These two look like the same customer"
                        : "\(pairs.count) pairs look like the same customer"
                )
                .font(.golos(13.5, weight: .semibold))
                .foregroundStyle(BrandColor.ink)
                Text("Merging keeps every message, task and photo from both, under one record.")
                    .font(.golos(11.5))
                    .foregroundStyle(BrandColor.muted500)
                    .fixedSize(horizontal: false, vertical: true)
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, 15)
        .padding(.vertical, 12)
    }

    private func row(_ pair: DuplicatePair) -> some View {
        HStack(alignment: .center, spacing: 10) {
            VStack(alignment: .leading, spacing: 2) {
                Text(
                    describeContact(pair.name_a, pair.phone_a)
                        + " and "
                        + describeContact(pair.name_b, pair.phone_b)
                )
                .font(.golos(13))
                .foregroundStyle(BrandColor.ink)
                .lineLimit(2)
                // The reason, in the words the server used. A suggestion
                // somebody cannot verify is one they learn to dismiss.
                Text(pair.reason)
                    .font(.golos(11))
                    .foregroundStyle(BrandColor.muted500)
                    .lineLimit(2)
            }
            .frame(maxWidth: .infinity, alignment: .leading)
            if canMerge {
                Button("Merge") { merging = pair }
                    .buttonStyle(.plain)
                    .font(.golos(11, weight: .semibold))
                    .foregroundStyle(BrandColor.muted700)
                    .padding(.horizontal, 12)
                    .padding(.vertical, 7)
                    .background(BrandColor.insetDeep, in: Capsule())
                    .accessibilityLabel(
                        "Merge \(describeContact(pair.name_a, pair.phone_a)) and "
                            + describeContact(pair.name_b, pair.phone_b)
                    )
            }
        }
        .padding(.horizontal, 15)
        .padding(.vertical, 11)
    }
}

/// A contact as somebody recognises it: the name if there is one, else the number.
func describeContact(_ name: String?, _ phone: String) -> String {
    let trimmed = name?.trimmingCharacters(in: .whitespacesAndNewlines)
    guard let trimmed, !trimmed.isEmpty else { return formatPhone(phone) }
    return "\(trimmed) (\(formatPhone(phone)))"
}

/// # Ethical Friction, and which direction the sheet states
///
/// A merge moves somebody's whole history under a different record. The undo
/// restores the second contact but NOT which thread came from which, so this
/// says out loud what survives and names the direction in the way people get
/// backwards ("merge A into B" is ambiguous to almost everyone).
///
/// Both numbers keep working either way — the fact that makes the decision safe
/// and the one most likely to be assumed wrong.
///
/// *Applying: Ethical Friction — a destructive action gets a deliberate,
/// stated confirmation. Smart Defaults — the survivor is preselected, so this
/// is never an unanswered form.*
@MainActor
private struct MergeContactsSheet: View {
    let mutations: ContactMutations
    let companyId: String
    let pair: DuplicatePair
    let onMerged: @MainActor (ContactMergeResult) -> Void

    @State private var keepFirst = true
    @State private var saving = false
    @State private var error: String?
    @Environment(\.dismiss) private var dismiss

    private var survivorId: String { keepFirst ? pair.contact_a : pair.contact_b }
    private var foldedId: String { keepFirst ? pair.contact_b : pair.contact_a }
    private var survivorLabel: String {
        keepFirst
            ? describeContact(pair.name_a, pair.phone_a)
            : describeContact(pair.name_b, pair.phone_b)
    }
    private var foldedLabel: String {
        keepFirst
            ? describeContact(pair.name_b, pair.phone_b)
            : describeContact(pair.name_a, pair.phone_a)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 10) {
            Text("Merge these two customers")
                .font(.golos(15, weight: .semibold))
                .foregroundStyle(BrandColor.ink)
            Text(
                "Everything from both — messages, tasks, photos, notes — ends up "
                    + "under the record you keep. Both phone numbers keep working."
            )
            .font(.golos(12.5))
            .foregroundStyle(BrandColor.muted600)
            .fixedSize(horizontal: false, vertical: true)

            Text("Which one to keep")
                .font(.golos(11.5, weight: .semibold))
                .foregroundStyle(BrandColor.muted700)
                .padding(.top, 2)

            ForEach([true, false], id: \.self) { first in
                let label = first
                    ? describeContact(pair.name_a, pair.phone_a)
                    : describeContact(pair.name_b, pair.phone_b)
                Button {
                    keepFirst = first
                } label: {
                    HStack(spacing: 9) {
                        Image(
                            systemName: keepFirst == first
                                ? "largecircle.fill.circle"
                                : "circle"
                        )
                        .font(.system(size: 16, weight: .regular))
                        .foregroundStyle(
                            keepFirst == first ? BrandColor.olive : BrandColor.muted300
                        )
                        Text(label)
                            .font(.golos(13))
                            .foregroundStyle(BrandColor.ink)
                            .multilineTextAlignment(.leading)
                        Spacer(minLength: 0)
                    }
                    .contentShape(Rectangle())
                }
                .buttonStyle(.plain)
                .disabled(saving)
                .accessibilityAddTraits(keepFirst == first ? [.isSelected] : [])
            }

            // Said back in the direction people get backwards.
            Text("\(foldedLabel) stops being a separate customer. Its history moves to \(survivorLabel).")
                .font(.golos(11.5))
                .foregroundStyle(BrandColor.muted500)
                .fixedSize(horizontal: false, vertical: true)
                .padding(.top, 2)

            if let error {
                Text(error)
                    .font(.golos(11.5))
                    .foregroundStyle(BrandColor.destructive)
                    .fixedSize(horizontal: false, vertical: true)
            }

            HStack(spacing: 10) {
                Spacer()
                Button("Cancel") { dismiss() }
                    .buttonStyle(.plain)
                    .font(.golos(12, weight: .semibold))
                    .foregroundStyle(BrandColor.muted600)
                    .disabled(saving)
                Button(saving ? "Merging…" : "Merge") { merge() }
                    .buttonStyle(.plain)
                    .font(.golos(12, weight: .semibold))
                    .foregroundStyle(BrandColor.paper)
                    .padding(.horizontal, 16)
                    .padding(.vertical, 9)
                    .background(BrandColor.ink, in: Capsule())
                    .disabled(saving)
            }
            .padding(.top, 4)
        }
        .padding(16)
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(BrandColor.canvas.ignoresSafeArea())
        .presentationDetents([.medium])
    }

    private func merge() {
        saving = true
        error = nil
        Task {
            do {
                let result = try await mutations.merge(
                    companyId: companyId,
                    fromContactId: foldedId,
                    intoContactId: survivorId
                )
                onMerged(result)
            } catch {
                self.error = error.userMessage
            }
            saving = false
        }
    }
}

// MARK: - Previews

#Preview("Duplicate contacts card") {
    let graph = AppGraph()
    ScrollView {
        DuplicateContactsCard(
            mutations: ContactMutations(
                api: graph.api,
                multipart: MultipartClient(api: graph.api, sessionStore: graph.sessionStore)
            ),
            companyId: "preview-co",
            canMerge: true,
            onMerged: { _ in }
        )
    }
    .background(BrandColor.canvas)
}
